### PR TITLE
Wrap expression str in Expression class, filter ANN only when needed, store ann_key in Expression

### PR DESCRIPTION
**Pull request by tedil**

### CODE REVIEW

File: vembrane/__init__.py [hunk 1]
```diff
@@ -108,31 +108,41 @@ def parse_annotation_entry(entry: str,) -> List[str]:
...
-                zip(annotation_keys, parse_annotation_entry(annotation)),
-            )
-        ),
-    )
-    return eval(expression, globals_whitelist, env)
+class Expression:
+    def __init__(self, expression: str, ann_key: str = "ANN"):
+        self._expression = expression
+        self._ann_key = ann_key
+        self._has_ann = f"{ann_key}[" in expression
```
Comment (tedil) [inline vembrane/__init__.py:115]: atm we just check for the ann_key with "[" appended to determine if the expression actually contains an ANN[ access. However, someone might do something like `"foo" in "".join(ANN)` for whatever reason, in which case it'd fail

File: vembrane/__init__.py [hunk 2]
```diff
@@ -108,31 +108,41 @@ def parse_annotation_entry(entry: str,) -> List[str]:
...
+
+    def annotation_key(self):
+        return self._ann_key
+
+    def filters_annotations(self):
+        return self._has_ann
+
+    def evaluate(
```
Comment (tedil) [inline vembrane/__init__.py:123]: this is just the eval_expression function moved into Expression and renamed

File: vembrane/__init__.py [hunk 3]
```diff
@@ -162,12 +172,18 @@ def filter_vcf(
...
+            filtered_annotations = [
+                annotation
+                for annotation in annotations
+                if expression.evaluate(idx, annotation, annotation_keys, env,)
+            ]
+        else:
+            annotations = [""]
+            filtered_annotations = (
+                [""] if expression.evaluate(idx, "", annotation_keys, env,) else []
+            )
```
Comment (tedil) [inline vembrane/__init__.py:186]: if there's no usage of ann_key in the expression, we do not need to re-evaluate the expression for every single annotation in the record but only need to do that exactly once

File: vembrane/__init__.py [hunk 4]
```diff
@@ -162,12 +172,18 @@ def filter_vcf(
...
         env["SAMPLES"] = list(record.samples)
 
-        annotations = dict(record.info).get(ann_key, [""])
-        filtered_annotations = [
-            annotation
-            for annotation in annotations
-            if eval_expression(expression, idx, annotation, annotation_keys, env,)
-        ]
+        if expression.filters_annotations():
+            annotations = dict(record.info).get(ann_key, [""])
```
Comment (FelixMoelder) [inline vembrane/__init__.py:176]: Do we need this empty string here? Should not be an empty list enough?

File: vembrane/__init__.py [hunk 5]
```diff
@@ -162,12 +172,18 @@ def filter_vcf(
...
+            filtered_annotations = [
+                annotation
+                for annotation in annotations
+                if expression.evaluate(idx, annotation, annotation_keys, env,)
+            ]
+        else:
+            annotations = [""]
```
Comment (FelixMoelder) [inline vembrane/__init__.py:183]: `annotations` gets initialized but is never used.

File: vembrane/__init__.py (hunk 4 above)
Comment (FelixMoelder) [inline vembrane/__init__.py:176]:
```suggestion
            annotations = dict(record.info).get(ann_key, [])
```

File: vembrane/__init__.py (hunk 5 above)
Comment (FelixMoelder) [inline vembrane/__init__.py:183]:
```suggestion
```

Comment (tedil) [reply vembrane/__init__.py:183]: it does get used further down!

File: vembrane/__init__.py (hunk 4 above)
Comment (tedil) [reply vembrane/__init__.py:176]: Yes, if there's an ann_key in the expression but there's no ANN field in the record, the list comprehension still needs to run at least once

Comment (tedil) [reply vembrane/__init__.py:176]: But I guess we can restructure code logic here so it isn't as confusing

Comment (FelixMoelder) [reply vembrane/__init__.py:176]: > Yes, if there's an ann_key in the expression but there's no ANN field in the record, the list comprehension still needs to have at least one entry (i.e. expression.evaluate needs to be called once, which would not happen if the list was empty)

Indeed! :)

**Pull request merged**